migrations/src/main/scala/za/co/absa/enceladus/migrations/continuous/migrate01/DatasetMigrator.scala

```diff
@@ -0,0 +1,157 @@
+      )
+    } match {
+      case Success(dataset1) =>
+        if (repo.doesDocumentExist(dataset1.name, dataset1.version)) {
+          resolveConflict(dataset1, objectId, repo)
+        } else {
+          normalInsert(dataset1, objectId, repo)
+        }
+      case Failure(e) =>
+        log.warn(s"Encountered a serialization error for '$collectionBase': ${e.getMessage}")
+    }
+  }
+
+  /**
+    * Migrates a conformance rule replacing a foreign key reference if necessary.
+    * Only mapping rules have foreign key references in the model version 0.
+    *
+    * @param rule A conformance rule from the model version 0.
+    * @return The transformed conformance rule.
+    */
+  def migrateConformanceRule(rule: model0.conformanceRule.ConformanceRule): model1.conformanceRule.ConformanceRule = {
+    val migratedConformanceRule = convertConformanceRule(rule)
+    migratedConformanceRule match {
+      case r: model1.conformanceRule.MappingConformanceRule =>
+        r.copy(mappingTableVersion = evm.getSafeVersion(EntityMigrator.mappingTableCollection,
+          r.mappingTable,
+          r.mappingTableVersion))
+      case r =>
+        r
+    }
+  }
+
+  /**
+    * In case there is no conflict, insert a new dataset normally.
+    * If inserting is unsuccessful fallback to conflict resolution.
+    *
+    * @param dataset  A dataset to save as an instance of Model 1 object
+    * @param objectId An Object Id if the dataset
+    * @param repo     An entity repository
+    */
+  def normalInsert(dataset: model1.Dataset, objectId: String, repo: EntityRepository): Unit = {
+    val dataset1Json = ObjectIdTools.putObjectIdIfNotPresent(Serializer1.serializeDataset(dataset),
+      Option(objectId))
+
+    try {
+      repo.insertDocument(dataset1Json)
+    } catch {
+      case NonFatal(e) =>
+        log.warn(s"Unable to append a document for '$collectionBase': ${e.getMessage}. Retrying...")
+        resolveConflict(dataset, objectId, repo)
+    }
+  }
+
+  /**
+    * In case there is a conflict, resolve it by adding a new version of the dataset
+    * and register the mapping between the old version and the new one in the entity version map.
+    *
+    * @param dataset  A dataset to save as an instance of Model 1 object
+    * @param objectId An Object Id if the dataset
+    * @param repo     An entity repository
+    */
+  def resolveConflict(dataset: model1.Dataset, objectId: String, repo: EntityRepository): Unit = {
+    var retriesLeft = EntityMigrator.NumberOfInsertRetries
+    var saved = false
+    while (retriesLeft > 0 && !saved) {
+      retriesLeft -= 1
+      val latestVersion = repo.getLatestVersion(dataset.name)
+      val datasetToSave = dataset.copy(version = latestVersion + 1)
+      val dataset1Json = ObjectIdTools
+        .putObjectIdIfNotPresent(Serializer1.serializeDataset(datasetToSave), Option(objectId))
+
+      try {
+        repo.insertDocument(dataset1Json)
```

Comment:
It can cause duplicates only if a document is inserted successfully, but a mapping cannot be added for some bizzare reason. Results of this will be a version of the document that is not used anywhere. Not a big deal for such a rare ephemeral event.